workflow/artifacts/s3/s3.go

```diff
@@ -57,30 +71,42 @@ func (s3Driver *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string)
 			s3cli, err := s3Driver.newS3Client(ctx)
 			if err != nil {
 				log.Warnf("Failed to create new S3 client: %v", err)
-				return false, nil
+				if isTransientS3Err(err) {
+					return false, nil
+				}
+				return false, fmt.Errorf("failed to create new S3 client: %v", err)
```

Comment:
Just to clarify, the `ExponentialBackoff` will stop and return if the condition function returns `true` or an error.

```go
type ConditionFunc func() (done bool, err error)
```

So we need to return:
1. `true, nil` if function finishes successfully
1. `false, fmt.Errorf("...")` if we want to return the error without retrying (for non-transient errors like `NoSuchBucket`)
1. `false, nil` if we want to retry (transient errors)

Reply:
You can just return `!isTransientS3Err(err), err`?

Reply:
Thanks @terrytangyuan! I'm a little confused here, please correct me if I'm wrong:

If we just return `!isTransientS3Err(err), err`:

1. If it's not a transient error, this will return `true, err`. `wait.ExponentialBackoff` will return immediately which is expected
2. If it is a transient error, this will return `false, err`. `wait.ExponentialBackoff` will return immediately too since `err != nil`. As a result, transient errors won't get retried.

For example, I tried the following:

```go
package main

import (
	"fmt"
	"time"

	"k8s.io/apimachinery/pkg/util/wait"
)

func main() {
	i := 0
	err := wait.ExponentialBackoff(wait.Backoff{Duration: time.Second * 2, Factor: 2.0, Steps: 5, Jitter: 0.1},
		func() (bool, error) {
			i++
			return false, fmt.Errorf("error")
		},
	)
	fmt.Println(err, i) // expecting output to be error, 1
}
```

Thanks!